lib/Language/Slicer/Eval.hs

```diff
@@ -19,178 +19,201 @@ import           Language.Slicer.UpperSemiLattice
-        ".svg" -> lift (visualizeDiffSVG s v1 v2) >> return VUnit
+        ".pdf" -> lift (visualizeDiffPDF s v1 v2) >> return (ORet VUnit)
+        ".svg" -> lift (visualizeDiffSVG s v1 v2) >> return (ORet VUnit)
         ext    -> evalError $ "visualizeDiff: unknown file extension : " ++ ext
-evalTraceOp PrimTreeSize [VTrace _ t _ _] =
-    return (VInt (forestsize (toTree t)))
-evalTraceOp PrimProfile [VTrace _ t _ _]
+evalTraceOp PrimTreeSize [ORet (VTrace _ t _ _)] =
+    return (ORet (VInt (forestsize (toTree t))))
+evalTraceOp PrimProfile [ORet (VTrace _ t _ _)]
     = do liftIO $ putStrLn (show (profile t))
-         return VUnit
-evalTraceOp PrimProfileDiff [VTrace _ t _ _]
+         return (ORet VUnit)
+evalTraceOp PrimProfileDiff [ORet (VTrace _ t _ _)]
     = do liftIO $ putStrLn (show (profileDiff t))
-         return VUnit
-evalTraceOp op vs = liftEvalM $ evalOp op vs
+         return (ORet VUnit)
+evalTraceOp op vs = liftEvalM $ evalOpExn op vs
+
+evalOpExn :: Primitive -> [Outcome] -> SlM Outcome
+evalOpExn f rs =
+  case extractExn rs of
+    Left vs -> do v <- evalOp f vs
+                  return (ORet v)
+    Right exn -> return (exn)
+  where extractExn [] = Left []
+        extractExn (ORet v : rs') = case extractExn rs' of
+                                        Left vs -> Left (v:vs)
+                                        Right exn -> Right exn
+        extractExn (OExn v : _) = Right (OExn v)
+        extractExn _ = Right (OHole)
 
 evalOp :: Primitive -> [Value] -> SlM Value
-evalOp f [VInt    i, VInt    j] | isCommonOp  f = return ((commonOps ! f) (i,j))
-evalOp f [VBool   i, VBool   j] | isCommonOp  f = return ((commonOps ! f) (i,j))
-evalOp f [VString i, VString j] | isCommonOp  f = return ((commonOps ! f) (i,j))
-evalOp f [VInt    i, VInt    j] | isIntBinOp  f = return ((intBinOps ! f) (i,j))
-evalOp f [VInt    i, VInt    j] | isIntRelOp  f = return ((intRelOps ! f) (i,j))
-evalOp f [VBool   i, VBool   j] | isBoolRelOp f = return ((boolRelOps! f) (i,j))
-evalOp f [VBool   b]            | isBoolUnOp  f = return ((boolUnOps ! f) b)
-evalOp _ vs                     | VHole `elem` vs = return VHole
-evalOp _ vs                     | VStar `elem` vs = return VStar
+evalOp f [(VInt    i), (VInt    j)] | isCommonOp  f = return ((commonOps ! f) (i,j))
+evalOp f [(VBool   i), (VBool   j)] | isCommonOp  f = return ((commonOps ! f) (i,j))
+evalOp f [(VString i), (VString j)] | isCommonOp  f = return ((commonOps ! f) (i,j))
+evalOp f [(VInt    i), (VInt    j)] | isIntBinOp  f = return ((intBinOps ! f) (i,j))
+evalOp f [(VInt    i), (VInt    j)] | isIntRelOp  f = return ((intRelOps ! f) (i,j))
+evalOp f [(VBool   i), (VBool   j)] | isBoolRelOp f = return ((boolRelOps! f) (i,j))
+evalOp f [(VBool   b)]            | isBoolUnOp  f = return ((boolUnOps ! f) b)
+evalOp _ vs                     | (VHole) `elem` vs = return (VHole)
+evalOp _ vs                     | (VStar) `elem` vs = return (VStar)
```

Comment:
Yes, I started to add `ORet` everywhere and later realized I didn't need it.  These can go.